Add LTO/IPO support.

ThisIsFineTM opened the pull request:
### Ticket
Closes #21844.

### Problem description
LTO was not supported in tt-metal cmake and had `-fno-lto`.

### What's changed
* Added project option TT_ENABLE_LTO
* Added cmake/linking.cmake
* linking.cmake attempts to enable LTO if TT_ENABLE_LTO is ON.
* If the checks in linking.cmake succeed, sets variable TT_LTO_ENABLED = TRUE.
* LTO being enabled also adds -Wodr for warning of One Definition Rule violations.
* References to -flto updated to -flto=auto to enable multithreaded LTO.
* LTO enabled by default in Release/RelWithDebInfo mode.
* Mold set to default linker.
* Mold hooked up to Clang LTO.
* LTO cache in build directory for faster incremental builds
* Debug sections now get compressed with zstd

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

ThisIsFineTM committed:
Add LTO/IPO support. #21844.

* Added project option TT_ENABLE_LTO (default: OFF if not Release)
* Added cmake/linking.cmake
* linking.cmake attempts to enable LTO if TT_ENABLE_LTO is ON.
* If the checks in linking.cmake succeed, set variable TT_LTO_ENABLED
  TRUE.
* LTO being enabled also adds -Wodr for warning of One Definition Rule
  violations.
* References to -flto updated to -flto=auto to enable multithreaded LTO.
* LTO enabled by default in Release/RelWithDebInfo mode.
* ODR violation warnings enabled.
* Mold set to default linker.
* Mold hooked up to Clang LTO.
* LTO cache in build directory for faster incremental builds
* Debug sections now get compressed with zstd

tt-rkim commented:
Turning it on by default, eh? Let's give it a go!

https://github.com/tenstorrent/tt-metal/actions/runs/14919856602

ThisIsFineTM commented:
> Turning it on by default, eh? Let's give it a go!
> 
> https://github.com/tenstorrent/tt-metal/actions/runs/14919856602

Its off by default. The flag in cmake/project_options.cmake controls the codepath in cmake/linking.cmake. I made two flags because cmake's test to see if IPO/LTO is available can fail.

tt-rkim commented:
Ah yes my apologies the first flag is `TT_ENABLE_LTO` - clearly illiterate.

ThisIsFineTM commented:
> Ah yes my apologies the first flag is `TT_ENABLE_LTO` - clearly illiterate.

That's not to say that having one of the CI runs flipping the switch would be a good idea. I'll do a cacheless benchmark to see how much time is added to the build. Now if you'll excuse me, I have to see about an owl for a little bit.

afuller-TT commented on METALIUM_GUIDE.md:
```diff
@@ -163,7 +163,7 @@ compilation command information like:
 && \
 /path/to/tt-metal//tt_metal/third_party/sfpi/compiler/bin/riscv32-unknown-elf-g++ \
--mgrayskull -march=rv32iy -mtune=rvtt-b1 -mabi=ilp32 -std=c++17 -flto -ffast-math \
+-mgrayskull -march=rv32iy -mtune=rvtt-b1 -mabi=ilp32 -std=c++17 -flto=auto -ffast-math \
```
(unrelated) grayskull is dead and gone; we should purge it from docs to not mislead ppl.

afuller-TT commented on cmake/linking.cmake:
```diff
@@ -0,0 +1,22 @@
+if(TT_ENABLE_LTO)
+    cmake_policy(SET CMP0069 NEW)
```
[CMP0069 was introduced in v3.9](https://cmake.org/cmake/help/latest/policy/CMP0069.html).
Our CMake is [setting all policies up to v3.30 ](https://github.com/tenstorrent/tt-metal/blob/main/CMakeLists.txt#L1)to NEW.
So this line is redundant.

```suggestion
```

afuller-TT commented on cmake/linking.cmake:
```diff
@@ -0,0 +1,22 @@
+if(TT_ENABLE_LTO)
+    cmake_policy(SET CMP0069 NEW)
+    set(CMAKE_POLICY_DEFAULT_CMP0069 NEW)
```
[CMP0069 was introduced in v3.9](https://cmake.org/cmake/help/latest/policy/CMP0069.html).
Our CMake is [setting all policies up to v3.30 ](https://github.com/tenstorrent/tt-metal/blob/main/CMakeLists.txt#L1)to NEW.
So these lines are redundant.

```suggestion
```

afuller-TT replied:
Update: Oh I see, it's because we have deps that have crazy old CMake min versions.  Like GTest on v3.5.  Those are issuing warnings.

ThisIsFineTM commented:
> > LTO being enabled also adds -Wodr for warning of One Definition Rule violations.
> 
> Are these related or orthogonal? Any reason we wouldn't just have -Wodr on all the time (aside from the raised issue that we currently violate the ODR 🙃 ) Or does -Wodr require LTO to detect the violation?

They are related. -Wodr requires LTO.

afuller-TT commented on tt_metal/hw/CMakeLists.txt:
```diff
@@ -153,7 +153,7 @@ set(GPP_FLAGS_blackhole -mcpu=tt-bh)
 set(GPP_FLAGS_common
     -std=c++17
-    -flto
+    -flto=auto
```
According to the [Clang docs](https://clang.llvm.org/docs/ClangCommandLineReference.html):
> -flto=<arg>, -flto (equivalent to -flto=full), -flto=auto (equivalent to -flto=full), -flto=jobserver (equivalent to -flto=full)[¶](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-flto)
Set LTO mode. <arg> must be ‘thin’ or ‘full’.

seems to say that `-flto` and `-flto=auto` are identical.
OTOH: `-flto=thin` looks like a more scalable choice: https://clang.llvm.org/docs/ThinLTO.html

I'll throw up a branch to see approx times in CI.  Assuming that lto=thin is the winnder, then I'd like to see it structured like this:

Option to opt-in/opt-out ✅ 
Detection using the module and setting TT_LTO_ENABLED ✅ 
top-level CMake after including linking.cmake and before processing any targets:
```cmake
set(CMAKE_INTERPROCEDURAL_OPTIMIZATION ${TT_LTO_ENABLED})
```
Delete all hardcoding of -flto / -fno-lto / etc

BRB gonna to try it out.

afuller-TT replied:
My bad.  My step 3 is already done.  But it's done inside linking.cmake instead of the top-level.  Not sure how I feel about that.  It's grouped by topic.  OTOH it means "what's being set in this project" is scattered.  Like if CMAKE_POSITION_INDEPENDENT_CODE were set in some other file.  Let me ponder a moment on that.

The flag, though, is setting CMAKE_IPO, so this PR will get -flto=thin for all that goodness.  I'm uncertain how much of the compile-time notes will get cached via ccache.  Hopefully it Just Works for the max incremental benefit that ThinLTO offers.

The flag here and in build.cpp are actually for the device, not host code.  And that's a different compiler than the host build that we're (mostly) looking at here.  AFAIK it doesn't even have ThinLTO.

afuller-TT replied:
Okay, here's my current thinking:

-flto -> -flto=auto seems to work (the CI was happy on this PR), so that's great.  If we're lucky we may even see our test runs be a little bit faster 🤞 

top-level option flag is perfect.

I'd like to reshuffle the linking.cmake.
I really value having all the settings in as flat a list as possible right in the root of the repo.  Ideally the list doesn't grow too large.  By having them as flat as possible it lets us grasp the state of affairs with as little cyclomatic complexity as possible.  And that in turn allows us to chip away at it to incrementally make it better.  (eg: fixing the disabled warnings one-by-one -- far easier when there's a single source of truth).
To that end, I would like the linking.cmake to just produce "Shall LTO be used?"
`CMAKE_INTERPROCEDURAL_OPTIMIZATION` I would like set along with our other global cmake flags that affect the build: https://github.com/tenstorrent/tt-metal/blob/main/CMakeLists.txt#L63
Again, as flat as possible, so rather than an `if` we can just set it to the value of "are we using LTO?"
In a similar vein, -Wodr I would like to see with all the other build flags: https://github.com/tenstorrent/tt-metal/blob/main/CMakeLists.txt#L168 just wrap it in a GenEx which keeps it flat and straightforward.

And since I misled you with removing the POLICY, as this will be affecting our vendor'd dependencies, some of which are still claiming an ancient version of CMake, that actually needs to go back in 🙈 

I'm not clear as to why we'd want `SEND_ERROR` (in honesty, I can't recall ever finding a use for "Spit out an error message, but then keep going and generate more haystack before finally dying at the end".  Just seems more difficult to find the needle.  If you find value in this I'd love to learn.

If not, then if we're doing a FATAL_ERROR anyway, we could just let check_ipo_supported do the FATAL_ERROR for us and save ourselves more if/else complexity.

I think it would all boil down to:
```cmake
set(TT_LTO_ENABLED OFF)
if(TT_ENABLE_LTO)
    include(CheckIPOSupported)
    check_ipo_supported() # Will error out if not supported
    set(TT_LTO_ENABLED ON)
endif()
```

afuller-TT replied:
Oh, and then we should enable it on the Release build.  Both to vet it and keep it from regressing, but also because don't we want our releases to be as fast as possible?  That's the whole point!  🙃

ThisIsFineTM replied:
@afuller-TT on clang, cmake's IPO will enable -flto=thin by default. Using -flto=auto also plays nice with gcc:

https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html
> If you specify the optional n, the optimization and code generation done at link time is executed in parallel using n parallel jobs by utilizing an installed make program. The environment variable MAKE may be used to override the program used.

> You can also specify -flto=jobserver to use GNU make’s job server mode to determine the number of parallel jobs. This is useful when the Makefile calling GCC is already executing in parallel. You must prepend a ‘+’ to the command recipe in the parent Makefile for this to work. This option likely only works if MAKE is GNU make. Even without the option value, GCC tries to automatically detect a running GNU make’s job server.

> Use -flto=auto to use GNU make’s job server, if available, or otherwise fall back to autodetection of the number of CPU threads present in your system.

ThisIsFineTM replied:
LTO can increase build times, but this is what you get for it:

* Inlining/optimization across translation units
* Deduplication, which also contributes to smaller file sizes
* The ability to catch ODR violations

ThisIsFineTM replied:
Re: incremental builds, we'll have to implement the additional caching flag.  Also, one thing I'd like to do is add either a default or another flag to enable using Mold as the linker by default, rather than just in toolchain files. 

https://clang.llvm.org/docs/ThinLTO.html#id11
> Incremental 
ThinLTO supports fast incremental builds through the use of a cache, which currently must be enabled through a linker option.
    gold (as of LLVM 4.0): -Wl,-plugin-opt,cache-dir=/path/to/cache
    ld64 (supported since clang 3.9 and Xcode 8) and Mach-O ld64.lld (as of LLVM 15.0): -Wl,-cache_path_lto,/path/to/cache
    ELF ld.lld (as of LLVM 5.0): -Wl,--thinlto-cache-dir=/path/to/cache
    COFF lld-link (as of LLVM 6.0): /lldltocache:/path/to/cache

ThisIsFineTM replied:
@afuller-TT 
Did some digging and found this thread:
https://github.com/rui314/mold/issues/181#issuecomment-1053434707
> I build mold with commit https://github.com/rui314/mold/commit/4caadefac64b99157231e3bb727c2674f8a5d328 now and could use, after this -Wl,--thinlto-jobs=all (LLVM/CLANG) to builds apps without problem. THANK YOU!

Also this:
> Let's close this, the full support will be available since GCC 13.1.

Looks like support for GCC 13+ is rock solid. Pre-GCC 13 also works but it needed a bit of a hack on the mold side of things, which doesn't look like it would affect the end user, but still something to be aware of. I'll take a look to see if mold has a cache dir flag like the others.

ThisIsFineTM replied:
@afuller-TT Would you be opposed to having `mold` as the default linker if its available rather than just having it used when a specific toolchain is set? Maybe something like:

```
if(NOT DEFINED CMAKE_LINKER_TYPE)
  find_program(MOLD ld.mold)
  if(MOLD)
    set(CMAKE_LINKER_TYPE MOLD)
    message(STATUS "Using mold linker.")
    if(TT_LTO_ENABLED)
      add_link_options($<$<CXX_COMPILER_ID:Clang>:-Wl,--thinlto-jobs=all>)
    endif()
  endif()
endif()

```

afuller-TT replied:
I'd support having an opinion on the defualt linker, yup!

Your code above would miss out on the --thinlto-jobs=all when using toolchain files, though.
Do we need to specify thinlto-jobs?  Is that not the default?
> By default, the ThinLTO link step will launch as many threads in parallel as there are cores.

https://clang.llvm.org/docs/ThinLTO.html#controlling-backend-parallelism

ThisIsFineTM replied:
@afuller-TT Mold isn't documented in that link, so thinlto probably doesn't know the specific flag to pass to mold. Notice how the arguments to the different linkers are different in that doc. I did some local tests, and mold wasn't using above 100% CPU when the flag was missing. I can propagate that to the clang toolchain file(s) to make sure we use it. Preferences?

ThisIsFineTM commented:
I added a flag to make ODR issues warn only for the time being since #21850 hasn't been addressed yet. Also did some more cleanup and added lto-cache support.

afuller-TT commented on CMakeLists.txt:
```diff
@@ -35,6 +35,11 @@ message(STATUS "Build type: ${CMAKE_BUILD_TYPE}")
 set(CMAKE_CXX_FLAGS_DEBUG "-O0 -g -DDEBUG")
 set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-O3 -g -DDEBUG")
+
+if(CMAKE_BUILD_TYPE STREQUAL "Release")
```
This assumes that build type is known at cmake configure time.  Which is not the case for multi-config generators.  Let me see where this is used and whether we can use a genex instead so all bases are covered.

ThisIsFineTM commented on cmake/linking.cmake:
```diff
@@ -0,0 +1,44 @@
+set(TT_LTO_ENABLED OFF)
+
+if(TT_ENABLE_LTO)
```
Did some digging and apparently this is a thing:
https://cmake.org/cmake/help/latest/variable/CMAKE_INTERPROCEDURAL_OPTIMIZATION_CONFIG.html
`CMAKE_INTERPROCEDURAL_OPTIMIZATION_<CONFIG>`

Tested this, and it did work:
```cmake
set(CMAKE_INTERPROCEDURAL_OPTIMIZATION_RELEASE ON)
```

ThisIsFineTM replied:
@afuller-TT Made those adjustments and did some cleanup/tuning.

afuller-TT commented on cmake/linking.cmake:
```diff
@@ -0,0 +1,66 @@
+
+        # Enable one-definition-rule (ODR) warnings.
+        # Only works when LTO is enabled.
+        # FIXME: Forcing odr to warn until current errors are addressed. (#21850)
```
I'd like us to start adopting [google-readability-todo](https://clang.llvm.org/extra/clang-tidy/checks/google/readability-todo.html).  Not sure if it checks both TODOs and FIXMEs.
```suggestion
        # TODO(21850): Forcing odr to warn until current errors are addressed.
```

afuller-TT commented on cmake/linking.cmake:
```diff
@@ -0,0 +1,66 @@
+            -Wno-error=odr
+        )
+
+        # make sure the linker will thread LTO (required for Mold integration)
```
The builds on the PR should include one or two GCC builds.  Let me confirm they run on fork PRs...
